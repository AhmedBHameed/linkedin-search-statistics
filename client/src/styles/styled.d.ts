import 'styled-components';

/**
 * For more information whey we need this file @see https://styled-components.com/docs/api#create-a-declarations-file
 */
declare module 'styled-components' {
  export interface DefaultTheme {
    colors: {
      blue5: string;
      blue50: string;
      blue90: string;
      red: string;
      highlightText: string;
    };
  }
}
