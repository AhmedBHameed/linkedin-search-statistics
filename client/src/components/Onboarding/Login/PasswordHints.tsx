import React from 'react';
import styled from 'styled-components';

// color: ${(props) => props.theme.colors.red};
const List = styled.ul`
  padding: 1.5rem;
  margin: 0;
`;

const Item = styled.li`
  font-size: 1rem;
  color: ${(props) => props.theme.colors.red};
`;

const PasswordHints: React.FC = () => (
  <List data-testid="password-hints">
    <Item>1 character</Item>
    <Item>1 number</Item>
    <Item>with no space</Item>
    <Item>8 characters minimum long</Item>
  </List>
);

export default PasswordHints;
