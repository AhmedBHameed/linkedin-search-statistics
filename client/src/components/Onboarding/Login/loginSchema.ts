import {Joi, requiredEmail, requiredPassword} from '../../../util/validations';
import {LoginInput} from './models/LoginInput';

const loginSchema = Joi.object<LoginInput>({
  email: requiredEmail(),
  password: requiredPassword(),
});

export default loginSchema;
