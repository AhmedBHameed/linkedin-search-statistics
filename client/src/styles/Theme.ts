/**
 * The number 50 is the scale of the color according to Figma
 * @see https://www.figma.com/file/WCipxn54SiYJSVpZ3uTUW5/BiomeDx-Styleguide-und-Report-%234?node-id=13%3A140
 */

const highlightText = '#ffc069';

const lightTheme = {
  red: 'red',
  blue5: '#e7f2fe',
  blue50: '#047CB3',
  blue90: '#011e2f',
  highlightText,
};

const darkTheme = {
  red: '#ff7777',
  blue5: '#e7f2fe',
  blue50: '#047CB3',
  blue90: '#011e2f',
  highlightText,
};

export type ThemeProps = typeof darkTheme;
export {lightTheme, darkTheme};
