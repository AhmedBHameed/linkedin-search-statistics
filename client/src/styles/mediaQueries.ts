export const sm = 576;
export const md = 768;
export const lg = 992;
export const xl = 1200;
export const xxl = 1600;

const devices = {
  smMediaQuery: `(min-width: ${sm}px)`,
  mdMediaQuery: `(min-width: ${md}px)`,
  lgMediaQuery: `(min-width: ${lg}px)`,
  xlMediaQuery: `(min-width: ${xl}px)`,
  xxlMediaQuery: `(min-width: ${xxl}px)`,
};

export default devices;
