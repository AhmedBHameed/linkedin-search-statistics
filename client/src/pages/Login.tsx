import {KeyOutlined, LoginOutlined, UserOutlined} from '@ant-design/icons';
import {joiResolver} from '@hookform/resolvers/joi';
import {Card, Col, Row, Typography} from 'antd';
import React, {useCallback, useState} from 'react';
import {Controller, useForm} from 'react-hook-form';
import {Link} from 'react-router-dom';
import styled from 'styled-components';
import useNavigateToDashboard from '../components/Dashboard/hooks/navigateToDashboardHook';
import {FormControl, InputField, PasswordField} from '../components/Forms/Form';
import loginSchema from '../components/Onboarding/Login/loginSchema';
import {LoginInput} from '../components/Onboarding/Login/models/LoginInput';
import PasswordHints from '../components/Onboarding/PasswordHints';
import auth from '../components/Onboarding/util/auth';
import {
  FullWidthLoadingButton,
  OnboardingForm,
  StyledBrand,
  StyledRow,
} from '../components/shared/onboardingStyles';
import {mediumGathomFontStyle} from '../components/shared/sharedStyles';
import useToastMessage from '../components/ToastMessage/hooks/toastMessageHook';
import NAVIGATION_ROUTES from '../config/NavigationRoutes';
import {callTryCatch} from '../util/callTryCatch';

const LoginTitle = styled(Typography.Title)`
  ${mediumGathomFontStyle}
  margin: 2rem 0;
  margin-bottom: 2rem !important;
`;

const Login: React.FC = () => {
  const {goToDashboard} = useNavigateToDashboard();
  const {notifyError} = useToastMessage();
  const [loading, setLoading] = useState(false);

  const login = useCallback(
    async (formData: LoginInput) => {
      setLoading(true);
      const [error] = await callTryCatch(async () =>
        auth.login({
          password: formData.password,
          username: formData.email,
        })
      );

      setLoading(false);

      if (error) {
        // eslint-disable-next-line no-console
        console.error('<Login />', error);
        notifyError({
          message: 'Login failed',
          description: 'Invalid username or password. Please try again later',
        });

        return;
      }

      goToDashboard();
    },
    [goToDashboard, setLoading, notifyError]
  );

  const {
    control,
    formState: {errors},
    handleSubmit,
  } = useForm<LoginInput>({
    resolver: joiResolver(loginSchema),
    defaultValues: {
      email: '',
      password: '',
    },
  });

  const emailError = errors.email?.message;
  const passwordError = errors.password?.message;

  return (
    <StyledRow align="middle" justify="center">
      <Col lg={6} md={10} sm={12} xs={24}>
        <OnboardingForm onSubmit={handleSubmit(login)} testId="login-form">
          <Card>
            <Row align="middle" justify="center">
              <StyledBrand />
            </Row>

            <Row>
              <Col xs={24}>
                <LoginTitle level={5}>Login</LoginTitle>
              </Col>
            </Row>

            <Row>
              <Col xs={24}>
                <FormControl error={emailError} label="Email">
                  <Controller
                    control={control}
                    name="email"
                    render={({field: {value, onChange}}) => (
                      <InputField
                        ariaLabel="email"
                        error={!!emailError}
                        name="email"
                        onChange={onChange}
                        placeholder="Email"
                        prefixIcon={<UserOutlined />}
                        size="large"
                        testId="email-input"
                        value={value}
                      />
                    )}
                  />
                </FormControl>
              </Col>

              <Col xs={24}>
                <FormControl error={passwordError} label="Password">
                  <Controller
                    control={control}
                    name="password"
                    render={({field: {value, onChange}}) => (
                      <PasswordField
                        ariaLabel="password"
                        error={!!passwordError}
                        name="password"
                        onChange={onChange}
                        placeholder="Password"
                        prefixIcon={<KeyOutlined />}
                        size="large"
                        testId="password-input"
                        value={value}
                      />
                    )}
                  />
                </FormControl>

                <Col xs={24}>
                  {errors.password?.message && <PasswordHints />}
                </Col>
              </Col>
            </Row>

            <Row justify="end">
              <Link to={NAVIGATION_ROUTES.forgotPassword.path}>
                Forgot password?
              </Link>
            </Row>

            <Row>
              <FullWidthLoadingButton
                disabled={loading}
                htmlType="submit"
                icon={<LoginOutlined />}
                loading={loading}
                size="large"
                testId="login-action-button"
              >
                Login
              </FullWidthLoadingButton>
            </Row>

            <Row align="middle" justify="center">
              <Link to={NAVIGATION_ROUTES.signup.path}>
                Register new account.
              </Link>
            </Row>
          </Card>
        </OnboardingForm>
      </Col>
    </StyledRow>
  );
};

export default Login;
