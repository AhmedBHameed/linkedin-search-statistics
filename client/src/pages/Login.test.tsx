import {act, fireEvent, render} from '@testing-library/react';

import {axe} from 'jest-axe';
import {BrowserRouter as Router} from 'react-router-dom';
import {ThemeProvider} from 'styled-components';
import {mockConsoleErrorFun, mockMediaQuery} from '../setupTests';
import {md} from '../styles/mediaQueries';
import {lightTheme} from '../styles/Theme';
import {rest, server} from '../test/testServer';
import {mockLoginData} from '../test/utils/generate';
import Login from './Login';

mockConsoleErrorFun();

const MOCK_LOGIN_DATA = mockLoginData({password: 'Test@123456789'});

const mockGoToDashboard = jest.fn();
jest.mock(
  '../components/Dashboard/hooks/navigateToDashboardHook',
  () => () => ({
    goToDashboard: mockGoToDashboard,
  })
);

const mockNotifyError = jest.fn();
jest.mock('../components/ToastMessage/hooks/toastMessageHook', () => () => ({
  notifyError: mockNotifyError,
}));

const makeSuccessfulCall = () => {
  server.use(
    rest.post(`*/posts`, (_, res, ctx) =>
      res(ctx.delay(500), ctx.status(200), ctx.json({result: 'Mocked result'}))
    )
  );
};

const makeErrorCall = () => {
  server.use(
    rest.post(`*/posts`, (_, res, ctx) =>
      res(ctx.status(500), ctx.json({error: 'async error'}))
    )
  );
};

function renderComponent() {
  mockMediaQuery(md, md);

  return render(
    <ThemeProvider theme={{colors: lightTheme}}>
      <Router>
        <Login />
      </Router>
    </ThemeProvider>
  );
}

describe('Login', () => {
  afterAll(() => {
    jest.clearAllMocks();
  });

  it('matches snapshot', () => {
    const {container} = renderComponent();

    expect(container).toMatchSnapshot();
  });

  describe('Form validation', () => {
    it('shows helper text as a hint when submitting invalid data.', async () => {
      const {findAllByTestId, getByTestId} = renderComponent();

      act(() => {
        fireEvent.submit(getByTestId('login-form'));
      });

      const inputErrors = await findAllByTestId('input-error');

      /**
       * The idea here is to match localization key of the error.
       */
      expect(inputErrors[0].textContent).toBe('validationError.required');
      expect(inputErrors[1].textContent).toBe('validationError.required');
      expect(mockGoToDashboard).toHaveBeenCalledTimes(0);
      expect(mockNotifyError).toHaveBeenCalledTimes(0);
    });

    it('submits valid data and redirect to dashboard.', async () => {
      makeSuccessfulCall();

      const {getByTestId} = renderComponent();

      /**
       * In case you wonder why we should use async act @see https://kentcdodds.com/blog/fix-the-not-wrapped-in-act-warning
       * Regarding line await waitFor(() => {})
       * It is a simulation for user behaviour (The test will fail without it).
       */
      fireEvent.change(getByTestId('email-input'), {
        target: {value: MOCK_LOGIN_DATA.email},
      });
      fireEvent.change(getByTestId('password-input'), {
        target: {value: MOCK_LOGIN_DATA.password},
      });
      fireEvent.submit(getByTestId('login-form'));

      /**
       * There is something wrong with react-testing-library
       * It is explained here @see https://github.com/testing-library/react-testing-library/issues/865
       * We are going to make a workaround till we find another better solution.
       */
      await new Promise((resolve) => setTimeout(resolve, 1000));

      expect(mockGoToDashboard).toHaveBeenCalledTimes(1);
      expect(mockGoToDashboard).toHaveBeenCalledWith(/* Nothing */);
    });

    it('triggers toast message when async throw an error', async () => {
      makeErrorCall();

      const {getByTestId} = renderComponent();

      /**
       * In case you wonder why we should use async act @see https://kentcdodds.com/blog/fix-the-not-wrapped-in-act-warning
       * Regarding line await waitFor(() => {})
       * It is a simulation for user behaviour (The test will fail without it).
       */
      fireEvent.change(getByTestId('email-input'), {
        target: {value: MOCK_LOGIN_DATA.email},
      });
      fireEvent.change(getByTestId('password-input'), {
        target: {value: MOCK_LOGIN_DATA.password},
      });
      fireEvent.click(getByTestId('login-action-button'));

      await new Promise((resolve) => setTimeout(resolve, 1000));

      expect(mockGoToDashboard).toHaveBeenCalledTimes(0);

      expect(mockNotifyError).toHaveBeenCalledTimes(1);
      expect(mockNotifyError).toHaveBeenCalledWith({
        message: 'login.error.invalidEmailOrPassword.title',
        description: 'login.error.invalidEmailOrPassword.message',
      });
    });
  });

  it('accessability check', async () => {
    makeSuccessfulCall();

    const {container} = renderComponent();

    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });
});
