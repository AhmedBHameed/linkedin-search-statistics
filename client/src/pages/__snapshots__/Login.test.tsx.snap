// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login matches snapshot 1`] = `
<div>
  <div
    class="ant-row ant-row-center ant-row-middle sc-kEqXSa fUiPLR"
  >
    <div
      class="ant-col ant-col-xs-24 ant-col-sm-12 ant-col-md-10 ant-col-lg-8"
    >
      <form
        class="sc-iqAclL kgxeLt"
        data-testid="login-form"
      >
        <div
          class="ant-card ant-card-bordered"
        >
          <div
            class="ant-card-body"
          >
            <div
              class="ant-row ant-row-center ant-row-middle"
            >
              <svg
                class="sc-bdnxRM goXlsq"
              >
                brand-light-theme.svg
              </svg>
            </div>
            <div
              class="ant-row"
            >
              <div
                class="ant-col ant-col-xs-24"
              >
                <h5
                  class="ant-typography sc-bqGGPW lgUozE"
                >
                  login.headTitle
                </h5>
              </div>
            </div>
            <div
              class="ant-row"
            >
              <div
                class="ant-col ant-col-xs-24"
              >
                <div
                  class="ant-row ant-form-item sc-eCApnc fExcEz"
                >
                  <div
                    class="ant-col ant-form-item-label"
                  >
                    <label
                      class=""
                      title="common.emailLabel"
                    >
                      common.emailLabel
                    </label>
                  </div>
                  <div
                    class="ant-col ant-form-item-control"
                  >
                    <div
                      class="ant-form-item-control-input"
                    >
                      <div
                        class="ant-form-item-control-input-content"
                      >
                        <span
                          class="ant-input-affix-wrapper ant-input-affix-wrapper-lg"
                        >
                          <span
                            class="ant-input-prefix"
                          >
                            <span
                              aria-label="user"
                              class="anticon anticon-user"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="user"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M858.5 763.6a374 374 0 00-80.6-119.5 375.63 375.63 0 00-119.5-80.6c-.4-.2-.8-.3-1.2-.5C719.5 518 760 444.7 760 362c0-137-111-248-248-248S264 225 264 362c0 82.7 40.5 156 102.8 201.1-.4.2-.8.3-1.2.5-44.8 18.9-85 46-119.5 80.6a375.63 375.63 0 00-80.6 119.5A371.7 371.7 0 00136 901.8a8 8 0 008 8.2h60c4.4 0 7.9-3.5 8-7.8 2-77.2 33-149.5 87.8-204.3 56.7-56.7 132-87.9 212.2-87.9s155.5 31.2 212.2 87.9C779 752.7 810 825 812 902.2c.1 4.4 3.6 7.8 8 7.8h60a8 8 0 008-8.2c-1-47.8-10.9-94.3-29.5-138.2zM512 534c-45.9 0-89.1-17.9-121.6-50.4S340 407.9 340 362c0-45.9 17.9-89.1 50.4-121.6S466.1 190 512 190s89.1 17.9 121.6 50.4S684 316.1 684 362c0 45.9-17.9 89.1-50.4 121.6S557.9 534 512 534z"
                                />
                              </svg>
                            </span>
                          </span>
                          <input
                            aria-label="email"
                            class="ant-input ant-input-lg"
                            data-testid="email-input"
                            name="email"
                            placeholder="common.emailPlaceholder"
                            type="text"
                            value=""
                          />
                        </span>
                        <small
                          class="sc-jSFjdj fmCvnz ant-form-item-explain ant-form-item-explain-error"
                          data-testid=""
                        >
                           
                        </small>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="ant-col ant-col-xs-24"
              >
                <div
                  class="ant-row ant-form-item sc-eCApnc fExcEz"
                >
                  <div
                    class="ant-col ant-form-item-label"
                  >
                    <label
                      class=""
                      title="common.passwordLabel"
                    >
                      common.passwordLabel
                    </label>
                  </div>
                  <div
                    class="ant-col ant-form-item-control"
                  >
                    <div
                      class="ant-form-item-control-input"
                    >
                      <div
                        class="ant-form-item-control-input-content"
                      >
                        <span
                          class="ant-input-affix-wrapper ant-input-affix-wrapper-lg ant-input-password ant-input-password-large"
                        >
                          <span
                            class="ant-input-prefix"
                          >
                            <span
                              aria-label="key"
                              class="anticon anticon-key"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="key"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M608 112c-167.9 0-304 136.1-304 304 0 70.3 23.9 135 63.9 186.5l-41.1 41.1-62.3-62.3a8.15 8.15 0 00-11.4 0l-39.8 39.8a8.15 8.15 0 000 11.4l62.3 62.3-44.9 44.9-62.3-62.3a8.15 8.15 0 00-11.4 0l-39.8 39.8a8.15 8.15 0 000 11.4l62.3 62.3-65.3 65.3a8.03 8.03 0 000 11.3l42.3 42.3c3.1 3.1 8.2 3.1 11.3 0l253.6-253.6A304.06 304.06 0 00608 720c167.9 0 304-136.1 304-304S775.9 112 608 112zm161.2 465.2C726.2 620.3 668.9 644 608 644c-60.9 0-118.2-23.7-161.2-66.8-43.1-43-66.8-100.3-66.8-161.2 0-60.9 23.7-118.2 66.8-161.2 43-43.1 100.3-66.8 161.2-66.8 60.9 0 118.2 23.7 161.2 66.8 43.1 43 66.8 100.3 66.8 161.2 0 60.9-23.7 118.2-66.8 161.2z"
                                />
                              </svg>
                            </span>
                          </span>
                          <input
                            action="click"
                            aria-label="password"
                            class="ant-input ant-input-lg"
                            data-testid="password-input"
                            name="password"
                            placeholder="common.passwordPlaceholder"
                            type="password"
                            value=""
                          />
                          <span
                            class="ant-input-suffix"
                          >
                            <span
                              aria-label="eye-invisible"
                              class="anticon anticon-eye-invisible ant-input-password-icon"
                              role="img"
                              tabindex="-1"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="eye-invisible"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M942.2 486.2Q889.47 375.11 816.7 305l-50.88 50.88C807.31 395.53 843.45 447.4 874.7 512 791.5 684.2 673.4 766 512 766q-72.67 0-133.87-22.38L323 798.75Q408 838 512 838q288.3 0 430.2-300.3a60.29 60.29 0 000-51.5zm-63.57-320.64L836 122.88a8 8 0 00-11.32 0L715.31 232.2Q624.86 186 512 186q-288.3 0-430.2 300.3a60.3 60.3 0 000 51.5q56.69 119.4 136.5 191.41L112.48 835a8 8 0 000 11.31L155.17 889a8 8 0 0011.31 0l712.15-712.12a8 8 0 000-11.32zM149.3 512C232.6 339.8 350.7 258 512 258c54.54 0 104.13 9.36 149.12 28.39l-70.3 70.3a176 176 0 00-238.13 238.13l-83.42 83.42C223.1 637.49 183.3 582.28 149.3 512zm246.7 0a112.11 112.11 0 01146.2-106.69L401.31 546.2A112 112 0 01396 512z"
                                />
                                <path
                                  d="M508 624c-3.46 0-6.87-.16-10.25-.47l-52.82 52.82a176.09 176.09 0 00227.42-227.42l-52.82 52.82c.31 3.38.47 6.79.47 10.25a111.94 111.94 0 01-112 112z"
                                />
                              </svg>
                            </span>
                          </span>
                        </span>
                        <small
                          class="sc-jSFjdj fmCvnz ant-form-item-explain ant-form-item-explain-error"
                          data-testid=""
                        >
                           
                        </small>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="ant-col ant-col-xs-24"
                />
              </div>
            </div>
            <div
              class="ant-row ant-row-end"
            >
              <a
                href="/password/forgotten"
              >
                login.forgotPasswordLabel
              </a>
            </div>
            <div
              class="ant-row"
            >
              <button
                class="ant-btn ant-btn-primary ant-btn-lg sc-dIsUp izwWel"
                data-testid="login-action-button"
                type="submit"
              >
                <span
                  aria-label="login"
                  class="anticon anticon-login"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="login"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <defs>
                      <style />
                    </defs>
                    <path
                      d="M521.7 82c-152.5-.4-286.7 78.5-363.4 197.7-3.4 5.3.4 12.3 6.7 12.3h70.3c4.8 0 9.3-2.1 12.3-5.8 7-8.5 14.5-16.7 22.4-24.5 32.6-32.5 70.5-58.1 112.7-75.9 43.6-18.4 90-27.8 137.9-27.8 47.9 0 94.3 9.3 137.9 27.8 42.2 17.8 80.1 43.4 112.7 75.9 32.6 32.5 58.1 70.4 76 112.5C865.7 417.8 875 464.1 875 512c0 47.9-9.4 94.2-27.8 137.8-17.8 42.1-43.4 80-76 112.5s-70.5 58.1-112.7 75.9A352.8 352.8 0 01520.6 866c-47.9 0-94.3-9.4-137.9-27.8A353.84 353.84 0 01270 762.3c-7.9-7.9-15.3-16.1-22.4-24.5-3-3.7-7.6-5.8-12.3-5.8H165c-6.3 0-10.2 7-6.7 12.3C234.9 863.2 368.5 942 520.6 942c236.2 0 428-190.1 430.4-425.6C953.4 277.1 761.3 82.6 521.7 82zM395.02 624v-76h-314c-4.4 0-8-3.6-8-8v-56c0-4.4 3.6-8 8-8h314v-76c0-6.7 7.8-10.5 13-6.3l141.9 112a8 8 0 010 12.6l-141.9 112c-5.2 4.1-13 .4-13-6.3z"
                    />
                  </svg>
                </span>
                <span
                  class="sc-jrsJWt gqswSH"
                >
                  login.actionButton
                </span>
              </button>
            </div>
            <div
              class="ant-row ant-row-center ant-row-middle"
            >
              <a
                href="/signup"
              >
                common.notRegistered
              </a>
            </div>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
